src/midi/midi_mt32.cpp

```diff
@@ -683,101 +694,124 @@ void MidiHandler_mt32::Close()
-		                                     last_played_frame * 2;
-		channel->AddSamples_sfloat(frames_to_be_played, sample_offset_in_buffer);
-		requested_frames -= frames_to_be_played;
-		last_played_frame += frames_to_be_played;
+	assert(channel);
+
+	constexpr auto warning_percent = 5;
+	if (const auto percent_full = frame_fifo.GetPercentFull(); percent_full < warning_percent) {
+		static auto iteration = 0;
+		if (iteration++ % 100 == 0) {
+			LOG_WARNING("MT32: Audio FIFO is %2d%% full", percent_full);
+		}
 	}
+
+	static std::vector<AudioFrame> frames = {};
+	frame_fifo.BulkDequeue(frames, requested_frames);
+	channel->AddSamples_sfloat(requested_frames, &frames[0][0]);
+
+	last_rendered_ms = PIC_FullIndex();
 }
 
-// Returns the number of frames left to play in the buffer.
-uint16_t MidiHandler_mt32::GetRemainingFrames()
+void MidiHandler_mt32::RenderFramesToFifo(const uint16_t num_frames)
 {
-	// If the current buffer has some frames left, then return those ...
-	if (last_played_frame < FRAMES_PER_BUFFER)
-		return FRAMES_PER_BUFFER - last_played_frame;
+	static std::vector<AudioFrame> frames = {};
+	// Maybe expand the vector
+	if (frames.size() < num_frames) {
+		frames.resize(num_frames);
+	}
 
-	// Otherwise put the spent buffer in backstock and get the next buffer
-	backstock.Enqueue(std::move(play_buffer));
-	play_buffer = playable.Dequeue();
-	total_buffers_played++;
-	last_played_frame = 0; // reset the frame counter to the beginning
+	std::unique_lock<std::mutex> lock(service_mutex);
+	service->renderFloat(&frames[0][0], num_frames);
+	lock.unlock();
 
-	return FRAMES_PER_BUFFER;
+	frame_fifo.BulkEnqueue(frames, num_frames);
 }
 
-// Keep the playable queue populated with freshly rendered buffers
-void MidiHandler_mt32::Render()
+// The next MIDI work task is processed, which includes rendering frames prior
+// to appling channel and sysex messages to the service
+void MidiHandler_mt32::ProcessWorkFromFifo()
 {
-	// Allocate our buffers once and reuse for the duration.
-	constexpr auto SAMPLES_PER_BUFFER = FRAMES_PER_BUFFER * 2; // L & R
-	std::vector<float> render_buffer(SAMPLES_PER_BUFFER);
-	std::vector<float> playable_buffer(SAMPLES_PER_BUFFER);
+	const auto work = work_fifo.Dequeue();
+
+	/* // Comment-in to log inter-cycle rendering
+	if (work.num_pending_frames > 0) {
+		LOG_MSG("MT32: %2u frames prior to %s message, followed by "
+		        "%2lu more messages. Have %4lu frames queued",
+		        work.num_pending_frames,
+		        work.message_type == MessageType::Channel ? "channel" : "sysex",
+		        work_fifo.Size(),
+		        frame_fifo.Size());
+	}*/
+
+	if (work.num_pending_frames > 0) {
+		RenderFramesToFifo(work.num_pending_frames);
+	}
 
-	// Populate the backstock using copies of the current buffer.
-	while (backstock.Size() < backstock.MaxCapacity() - 1)
-		backstock.Enqueue(playable_buffer);
+	// Request excuslive access prior to applying messages
```

Comment:
typo: exclusive

Reply:
Sounds an en excuse, specifically meant for you.  It's an excuslive.

Reply:
fixed :-)

Reply:
Haha 😂